### PR TITLE
Implement the functionality to just wait for the last event when dumping history events

### PR DESCRIPTION
**Pull request by wxing1292**
…vent of when dumping history events

### CODE REVIEW

File: service/frontend/handler.go [hunk 1]
```diff
@@ -1029,31 +1020,50 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
...
 	}
 
 	history := &gen.History{}
-	if token.FirstEventID >= token.NextEventID {
-		// currently there is no new event
-		events := []*gen.HistoryEvent{}
-		history.Events = events
-		if !token.IsWorkflowRunning {
+	history.Events = []*gen.HistoryEvent{}
+	if isCloseEventOnly {
```
Comment (samarabbas) [inline service/frontend/handler.go]: Probably it makes sense to push the filter all the way down to history service.

File: common/persistence/dataInterfaces.go [hunk 2]
```diff
@@ -127,6 +127,7 @@ type (
 		ExecutionContext             []byte
 		State                        int
 		CloseStatus                  int
+		LastFirstEventID             int64
```
Comment (samarabbas) [inline common/persistence/dataInterfaces.go]: You need persistence changes to read/write this new field.

File: service/frontend/handler.go [hunk 3]
```diff
@@ -1029,31 +1020,50 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
...
+		if !isWorkflowRunning {
+			history, _, err = wh.getHistory(domainInfo.ID, *execution, lastFirstEventID, nextEventID,
+				*getRequest.MaximumPageSize, nil, token.TransientDecision)
+			if err != nil {
+				return nil, wh.error(err, scope)
+			}
+			// since getHistory func will not return empty history, so the below is safe
+			history.Events = history.Events[len(history.Events)-1 : len(history.Events)]
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/frontend/handler.go]: https://github.com/uber/cadence/blob/master/common/persistence/cassandraHistoryPersistence.go#L180

if no events return (or empty), persistence will return err

File: service/frontend/handler.go (hunk 1 above)
Comment (wxing1292) [reply service/frontend/handler.go]: ok

File: service/frontend/handler.go [hunk 4]
```diff
@@ -960,41 +960,32 @@ func (wh *WorkflowHandler) GetWorkflowExecutionHistory(
...
-
-		if _, ok := err.(*gen.EntityNotExistsError); !ok || execution.RunId == nil {
-			return "", 0, false, err
-		}
-		// here we have error == EntityNotExistsError && run ID != nil
-		// (validateExecution guaranee that when run ID != nil, run ID is a valid UUID)
-
-		// It is possible that we still have the events in the table even though the mutable state is gone
-		// Get the nextEventID from visibility store if we still have it.
-		visibilityResp, err := wh.visibitiltyMgr.GetClosedWorkflowExecution(&persistence.GetClosedWorkflowExecutionRequest{
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/frontend/handler.go]: searching using vs code
the call GetClosedWorkflowExecution only appears in:
VisibilityManager.go
cassandraVisibilityPersistence_test.go
cassandraVisibilityPersistence.go
visibilityInterfaces.go

so i think we are good

File: idl/github.com/uber/cadence/history.thrift [hunk 5]
```diff
@@ -50,10 +50,12 @@ struct GetWorkflowExecutionNextEventIDRequest {
 }
 
 struct GetWorkflowExecutionNextEventIDResponse {
-  10: optional i64 (js.type = "Long") eventId
+  10: optional i64 (js.type = "Long") nextEventId
   20: optional string runId
   30: optional shared.TaskList tasklist
   40: optional bool isWorkflowRunning
+  // the event ID in the history event table which represent last batch of events
+  50: optional i64 (js.type = "Long") lastFirstEventId
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply idl/github.com/uber/cadence/history.thrift]: already done in https://github.com/uber/cadence/pull/464

File: service/history/historyEngine.go [hunk 6]
```diff
@@ -285,6 +285,7 @@ func (e *historyEngineImpl) StartWorkflowExecution(startRequest *h.StartWorkflow
 	if err1 != nil {
 		return nil, err1
 	}
+	msBuilder.executionInfo.LastFirstEventID = *startedEvent.EventId
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/historyEngine.go]: ok

File: service/frontend/handler.go (hunk 1 above)
Comment (wxing1292) [reply service/frontend/handler.go]: done using the int64 max as expected next event ID

File: common/persistence/dataInterfaces.go (hunk 2 above)
Comment (wxing1292) [reply common/persistence/dataInterfaces.go]: done

File: common/persistence/dataInterfaces.go [hunk 7]
```diff
@@ -360,6 +360,7 @@ type (
 		WorkflowTimeout             int32
 		DecisionTimeoutValue        int32
 		ExecutionContext            []byte
+		LastFirstEventID            int64
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply common/persistence/dataInterfaces.go]: done

**Pull request merged**